File: backend/app/schemas/experience.py
from pydantic import BaseModel, field_validator, field_serializer
from datetime import datetime, date, timezone
from typing import Optional, List
from app.models.experience import ExperienceCategory


class UserInfo(BaseModel):
    id: str
    username: str
    email: str
    
    @field_validator('id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        return str(v) if v is not None else v
    
    class Config:
        from_attributes = True


class ExperienceBase(BaseModel):
    title: str
    description: Optional[str] = None
    start_date: date
    end_date: Optional[date] = None
    is_current: bool = False
    category: Optional[ExperienceCategory] = None
    tags: Optional[List[str]] = []
    source_record_id: Optional[str] = None
    is_public: bool = False
    
    class Config:
        use_enum_values = True


class ExperienceCreate(ExperienceBase):
    pass


class ExperienceUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    is_current: Optional[bool] = None
    category: Optional[ExperienceCategory] = None
    tags: Optional[List[str]] = None
    is_public: Optional[bool] = None


class ExperienceResponse(ExperienceBase):
    id: str
    tenant_id: str
    user_id: str
    user: Optional[UserInfo] = None
    order_index: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    @field_serializer('created_at', 'updated_at')
    @classmethod
    def serialize_datetime(cls, dt: Optional[datetime]) -> Optional[str]:
        if dt is None:
            return None
        if dt.tzinfo is not None:
            dt = dt.astimezone(timezone.utc)
        return dt.strftime('%Y-%m-%dT%H:%M:%S.%f')[:-3] + 'Z'
    
    @field_validator('id', 'tenant_id', 'user_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v):
        return str(v) if v is not None else v
    
    class Config:
        from_attributes = True


class ExperienceListResponse(BaseModel):
    data: List[ExperienceResponse]
    total: int
    page: int
    page_size: int

File: frontend/src/types/experience.ts
export type ExperienceCategory = 'work' | 'project' | 'education' | 'milestone';

export interface UserInfo {
  id: string;
  username: string;
  email: string;
  avatar?: string;
}

export interface Experience {
  id: string;
  tenant_id: string;
  user_id: string;
  user?: UserInfo;
  title: string;
  description?: string;
  start_date: string;
  end_date?: string;
  is_current: boolean;
  category?: ExperienceCategory;
  tags: string[];
  source_record_id?: string;
  order_index: number;
  is_public: boolean;
  created_at: string;
  updated_at?: string;
  record_count?: number;
}

export interface ExperienceCreate {
  title: string;
  description?: string;
  start_date: string;
  end_date?: string;
  is_current?: boolean;
  category?: ExperienceCategory;
  tags?: string[];
  source_record_id?: string;
  is_public?: boolean;
}

export interface ExperienceUpdate {
  title?: string;
  description?: string;
  start_date?: string;
  end_date?: string;
  is_current?: boolean;
  category?: ExperienceCategory;
  tags?: string[];
  is_public?: boolean;
}

export interface ExperienceListResponse {
  data: Experience[];
  total: number;
  page: number;
  page_size: number;
}
